Jenkinsfile

```diff
@@ -11,48 +11,68 @@
-            timeout(time: 180, unit: 'MINUTES') {
-                // See below for what this method does - we're passing an arbitrary environment
-                // variable to it so that JAVA_OPTS and MAVEN_OPTS are set correctly.
-                withMavenEnv(["JAVA_OPTS=-Xmx1536m -Xms512m -XX:MaxPermSize=1024m",
-                            "MAVEN_OPTS=-Xmx1536m -Xms512m -XX:MaxPermSize=1024m"]) {
-                    // Actually run Maven!
-                    // The -Dmaven.repo.local=${pwd()}/.repository means that Maven will create a
-                    // .repository directory at the root of the build (which it gets from the
-                    // pwd() Workflow call) and use that for the local Maven repository.
-                    sh "mvn -Pdebug -U clean install ${runTests ? '-Dmaven.test.failure.ignore=true' : '-DskipTests'} -V -B -Dmaven.repo.local=${pwd()}/.repository"
+def builds = [:]
+for(i = 0; i < buildTypes.size(); i++) {
+    def buildType = buildTypes[i]
+    builds[buildType] = {
+        node(buildType.toLowerCase()) {
+            timestamps {
+                // First stage is actually checking out the source. Since we're using Multibranch
+                // currently, we can use "checkout scm".
+                stage('Checkout') {
+                    checkout scm
                 }
-            }
-        }
 
+                // Now run the actual build.
+                stage("${buildType} Build / Test") {
+                    timeout(time: 180, unit: 'MINUTES') {
+                        // See below for what this method does - we're passing an arbitrary environment
+                        // variable to it so that JAVA_OPTS and MAVEN_OPTS are set correctly.
+                        withMavenEnv(["JAVA_OPTS=-Xmx1536m -Xms512m -XX:MaxPermSize=1024m",
+                                    "MAVEN_OPTS=-Xmx1536m -Xms512m -XX:MaxPermSize=1024m"]) {
+                            // Actually run Maven!
+                            // The -Dmaven.repo.local=${pwd()}/.repository means that Maven will create a
+                            // .repository directory at the root of the build (which it gets from the
+                            // pwd() Workflow call) and use that for the local Maven repository.
+                            def mvnCmd = "mvn -Pdebug -U clean install ${runTests ? '-Dmaven.test.failure.ignore=true' : '-DskipTests'} -V -B -Dmaven.repo.local=${pwd()}/.repository" 
+                            if(isUnix()) {
+                                sh mvnCmd
+                            } else {
+                                bat mvnCmd
+                            }
+                        }
+                    }
+                }
+
+                // Once we've built, archive the artifacts and the test results.
+                stage("${buildType} Archive Artifacts / Test Results") {
+                    def files = findFiles(glob: '**/target/*.jar, **/target/*.war, **/target/*.hpi')
```

Comment:
I suppose _Pipeline Utility Steps_ is installed in ci.jenkins.io right? @daniel-beck @rtyler